In [274]:
import numpy as np
import pandas as pd
from scipy.special import expit
from imblearn.over_sampling import SMOTE

In [275]:
# I relied on this book: Tariq Rashid, Make your own neural network
# https://github.com/makeyourownneuralnetwork/makeyourownneuralnetwork/blob/master/part2_neural_network.ipynb

In [276]:
data=pd.read_csv('Roonyx/winequality-red.csv')

In [277]:
data.head(2)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70          0.0             1.9      0.076   
1            7.8              0.88          0.0             2.6      0.098   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   

   alcohol  quality  
0      9.4        5  
1      9.8        5

In [278]:
X=data.iloc[:,0:11].values
Y=data.loc[:,'quality'].values 

In [279]:
sm=SMOTE(random_state=3)

In [280]:
X, Y = sm.fit_sample(X, Y)

In [281]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(
    X, Y, test_size=0.2, random_state=42)

In [282]:
Y_train=Y_train-3 
Y_test=Y_test-3
#it means: 
#0 output node->quality 3
#1 node->quality 4 e.t.c

In [283]:
for i in range(11):
    X_test[:,i]/=np.max(X_test[:,i])
    X_test[:,i]+=0.01
X_test #normalize for sigmoid [0.01,1.0]

array([[0.82935484, 0.38341772, 0.51211338, ..., 0.77923077, 0.41217639,
        0.67428571],
       [0.50032258, 0.35177215, 0.15505498, ..., 0.87923077, 0.360468  ,
        0.68142857],
       [0.71039109, 0.26038366, 0.53442953, ..., 0.84759267, 0.45621635,
        0.8489415 ],
       ...,
       [0.46275443, 0.54882084, 0.05890253, ..., 0.88970894, 0.32176383,
        0.73489704],
       [0.43694142, 0.38804568, 0.01072895, ..., 0.93317983, 0.36471202,
        0.8290954 ],
       [0.48741935, 0.34544304, 0.30010995, ..., 0.81769231, 0.33748649,
        0.68142857]])

In [284]:
for i in range(11):
    X_train[:,i]/=np.max(X_train[:,i])
    X_train[:,i]+=0.01
X_train #normalize for sigmoid [0.01,1.0]

array([[0.4856124 , 0.98171562, 0.01377629, ..., 0.88093452, 0.21453155,
        0.73799543],
       [0.4372072 , 0.30834879, 0.16407562, ..., 0.85803441, 0.26240808,
        0.7510726 ],
       [0.52617619, 0.38969862, 0.24247952, ..., 0.80265829, 0.54564036,
        0.62463081],
       ...,
       [0.46283019, 0.40460421, 0.07      , ..., 0.88531172, 0.28      ,
        0.64758389],
       [0.6120111 , 0.2814778 , 0.46345555, ..., 0.80055942, 0.34910029,
        0.80503474],
       [0.66425622, 0.24444192, 0.50314811, ..., 0.78388834, 0.42383748,
        0.71635516]])

In [285]:
# X_train,X_test,Y_train,Y_test

In [286]:
framex=pd.DataFrame(X_train) #new train set
framex['quality']=Y_train
framex.head(4)

0         1         2         3         4         5         6  \
0  0.485612  0.981716  0.013776  0.163025  0.227424  0.077347  0.042449   
1  0.437207  0.308349  0.164076  0.133935  0.157603  0.065556  0.035460   
2  0.526176  0.389699  0.242480  0.115006  0.139604  0.376945  0.407716   
3  0.514063  0.637239  0.037135  0.264645  0.146701  0.083424  0.050531   

          7         8         9        10  quality  
0  1.001353  0.880935  0.214532  0.737995        0  
1  1.000728  0.858034  0.262408  0.751073        1  
2  1.003534  0.802658  0.545640  0.624631        1  
3  1.005048  0.873086  0.258567  0.735486        0

In [287]:
framex=framex.sample(frac=1)#mix it

In [288]:
framex.head(4)

0         1         2         3         4         5         6  \
3251  0.599491  0.230939  0.474581  0.153438  0.122917  0.148889  0.094762   
750   0.614360  0.496491  0.323038  0.177063  0.265850  0.084674  0.058849   
1573  0.575980  0.272077  0.496544  0.129927  0.126891  0.108390  0.065654   
1946  0.758428  0.379146  0.670000  0.171290  0.127840  0.093333  0.138028   

             7         8         9        10  quality  
3251  1.001959  0.818656  0.426317  0.808556        4  
750   1.005546  0.828076  0.298644  0.639459        0  
1573  1.000773  0.810289  0.356261  0.852313        5  
1946  1.005527  0.770599  0.290000  0.681141        2

In [289]:
framex=framex.reset_index(drop=True)#return indexes
framex.head(2)

0         1         2         3         4         5         6  \
0  0.599491  0.230939  0.474581  0.153438  0.122917  0.148889  0.094762   
1  0.614360  0.496491  0.323038  0.177063  0.265850  0.084674  0.058849   

          7         8         9        10  quality  
0  1.001959  0.818656  0.426317  0.808556        4  
1  1.005546  0.828076  0.298644  0.639459        0

In [290]:
X_new=framex.iloc[:,0:11].values
Y_new=framex.loc[:,'quality'].values #final train sets

In [291]:
#X_new,Y_new

In [292]:
class NeuralNetwork:
    
    def __init__(self,input_nodes,hidden_nodes,output_nodes,learning_rate):
        self.input=input_nodes
        self.hidden=hidden_nodes
        self.output=output_nodes
        self.lr=learning_rate
        self.wih=np.random.normal(0.0,self.input**-0.5,(self.hidden,self.input)) #w_input_hidden1 
        
        
        self.whh=np.random.normal(0.0,self.hidden**-0.5,(self.hidden,self.hidden)) #w_hidden1_hidden2
        
        self.who=np.random.normal(0.0,self.hidden**-0.5,(self.output,self.hidden)) #w_hidden2_output
       
    
        pass
        
    
    def train(self,X,Y):
        x=np.reshape(X,(11,1)) #1 object, 11 features
        y=np.reshape(Y,(6,1)) # 6 nodes for 6 classes
        hidden_inputs=np.dot(self.wih,x) # w_input_hidden1 * inputs
        hidden_outputs=expit(hidden_inputs) #sigmoid
        
        hidden_inputs2=np.dot(self.whh,hidden_outputs) #w_hidden1_hidden2 * hidden1_out
        hidden_outputs2=expit(hidden_inputs2)
        
        final_inputs=np.dot(self.who,hidden_outputs2)#w_hidden2_output * hidden2_out
        final_outputs=expit(final_inputs)
        
        output_errors=y-final_outputs #errors [6,1]
        
        
        hidden_errors2=np.dot(self.who.T,output_errors)# errors in hiiden2 nodes
        
        self.who+=2*self.lr*np.dot((output_errors*final_outputs * (1.0-final_outputs)),hidden_outputs2.T)
        #it`s derivative: 2*(error)*Ok(1-Ok)*Oj.T and lr -> learning rate
        
        self.whh+=2*self.lr*np.dot((hidden_errors2 * hidden_outputs2 * (1.0-hidden_outputs2)),hidden_outputs.T)
        
        hidden_errors=np.dot(self.whh.T,hidden_errors2)# errors in hidden1 nodes
        
        self.wih+=2*self.lr*np.dot((hidden_errors2 * hidden_outputs * (1.0-hidden_outputs)),x.T)
        
        pass
    
    
    def predict(self,X):
        x=np.reshape(X,(11,1))
        hidden_inputs=np.dot(self.wih,x)
        hidden_outputs=expit(hidden_inputs)
        
        hidden_inputs2=np.dot(self.whh,hidden_outputs)
        hidden_outputs2=expit(hidden_inputs2)
        
        final_inputs=np.dot(self.who,hidden_outputs2)
        final_outputs=expit(final_inputs)    
        return final_outputs
    
            
    
        
    

In [293]:
D8=NeuralNetwork(11,10,6,0.15)#11 input nodes, 3 hidden nodes in each one(6), 6 output nodes

In [300]:
%%time
#train
for e in range(5):#epochs
    i=0
    for one_object in X_new:#for every object 
        targets=np.zeros(6)+0.01 #target value of this object
        targets[Y_new[i]]=0.99 #for example target is 3 -> [0.99, 0.01, 0.01, 0.01, 0.01, 0.01]
        D8.train(one_object,targets)
        i+=1 #new target for new object

Wall time: 1.2 s


In [301]:
#predictions on the train set
i=0
score=[]
for test in X_new:
    targets=np.zeros(6)+0.01
    targets[Y_new[i]]=0.99
    correct_answer=Y_new[i]+3
    
    pred=D8.predict(test)
    label =np.argmax(pred)+3
    
    if (label==correct_answer): #if prediction is true append 1, else eappend 0
        score.append(1)
    else:
        score.append(0)
    i+=1
score=np.array(score)   
print('Accuracy=',score.sum()/score.size)
    

Accuracy= 0.5651774785801713


In [302]:
#predictions on the test set
i=0
score=[]
for test in X_test:
    targets=np.zeros(6)+0.01
    targets[Y_test[i]]=0.99
    correct_answer=Y_test[i]+3
    
    pred=D8.predict(test)
    label =np.argmax(pred)+3
    
    if (label==correct_answer): #if prediction is true append 1, else eappend 0
        score.append(1)
    else:
        score.append(0)
    i+=1
score=np.array(score)   
print('Accuracy=',score.sum()/score.size)

Accuracy= 0.5513447432762836


## Keras

In [210]:
import keras
from keras.utils import np_utils
Y_new = np_utils.to_categorical(Y_new,6)
Y_test = np_utils.to_categorical(Y_test,6)

In [211]:
from keras import models
from keras import layers
from keras import regularizers

In [271]:
network = models.Sequential()
network.add(layers.InputLayer([11])) #11 input nodes
network.add(layers.Dense(10, activation='sigmoid')) #hidden nodes
network.add(layers.Dense(10, activation='sigmoid')) #hidden nodes
network.add(layers.Dense(6, activation='softmax')) #output nodes

In [272]:
from keras import optimizers
sgd = optimizers.SGD(lr=0.15, decay=1e-6, momentum=0.9, nesterov=True)
network.compile(optimizer=sgd, loss='categorical_crossentropy',
                metrics=['accuracy'])

In [273]:
%%time
network.fit(X_new, Y_new, epochs=25, batch_size=10)

Epoch 1/25
3268/3268 [==============================] - 1s 181us/step - loss: 1.8225 - accuracy: 0.1631
Epoch 2/25
3268/3268 [==============================] - 0s 116us/step - loss: 1.4896 - accuracy: 0.3228
Epoch 3/25
3268/3268 [==============================] - 0s 113us/step - loss: 1.2574 - accuracy: 0.4113
Epoch 4/25
3268/3268 [==============================] - 0s 113us/step - loss: 1.2060 - accuracy: 0.4510
Epoch 5/25
3268/3268 [==============================] - 0s 114us/step - loss: 1.1797 - accuracy: 0.4584
Epoch 6/25
3268/3268 [==============================] - 0s 115us/step - loss: 1.1576 - accuracy: 0.4841
Epoch 7/25
3268/3268 [==============================] - 0s 114us/step - loss: 1.1381 - accuracy: 0.5003
Epoch 8/25
3268/3268 [==============================] - 0s 115us/step - loss: 1.1203 - accuracy: 0.5046
Epoch 9/25
3268/3268 [==============================] - 0s 113us/step - loss: 1.1101 - accuracy: 0.5144
Epoch 10/25
3268/3268 [==============================] - 0s 113u

In [263]:
test_loss, test_acc = network.evaluate(X_test, Y_test)

818/818 [==============================] - 0s 98us/step


In [264]:
test_acc

0.57334965467453

## Conclusion

1) NN on Pyton + Numpy:  
*You can go deeper in nn and fix it indise*  
*train time 6s for this dataset (25 epochs)*  
*test accuracy=0.55*  
*helps you to understand what is going on inside*  
*but it takes more time to make it than keras nn* 

2) Keras NN  
*train time 10s for 25 epochs, batch=10*  
*test accuracy = 0.573, higher than numpy_nn*  
*it takes a little time to do this nn with using keras*



